backend/pkg/database/gorm/asset.go

```diff
@@ -0,0 +1,316 @@
+}
+
+type AssetsTableHandler struct {
+	DB *gorm.DB
+}
+
+func (db *Handler) AssetsTable() types.AssetsTable {
+	return &AssetsTableHandler{
+		DB: db.DB,
+	}
+}
+
+func (t *AssetsTableHandler) GetAssets(params models.GetAssetsParams) (models.Assets, error) {
+	var assets []Asset
+	err := ODataQuery(t.DB, assetSchemaName, params.Filter, params.Select, params.Expand, params.OrderBy, params.Top, params.Skip, true, &assets)
+	if err != nil {
+		return models.Assets{}, err
+	}
+
+	items := make([]models.Asset, len(assets))
+	for i, tr := range assets {
+		var asset models.Asset
+		err = json.Unmarshal(tr.Data, &asset)
+		if err != nil {
+			return models.Assets{}, fmt.Errorf("failed to convert DB model to API model: %w", err)
+		}
+		items[i] = asset
+	}
+
+	output := models.Assets{Items: &items}
+
+	if params.Count != nil && *params.Count {
+		count, err := ODataCount(t.DB, assetSchemaName, params.Filter)
+		if err != nil {
+			return models.Assets{}, fmt.Errorf("failed to count records: %w", err)
+		}
+		output.Count = &count
+	}
+
+	return output, nil
+}
+
+func (t *AssetsTableHandler) GetAsset(assetID models.AssetID, params models.GetAssetsAssetIDParams) (models.Asset, error) {
+	var dbAsset Asset
+	filter := fmt.Sprintf("id eq '%s'", assetID)
+	err := ODataQuery(t.DB, assetSchemaName, &filter, params.Select, params.Expand, nil, nil, nil, false, &dbAsset)
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return models.Asset{}, types.ErrNotFound
+		}
+		return models.Asset{}, err
+	}
+
+	var apiAsset models.Asset
+	err = json.Unmarshal(dbAsset.Data, &apiAsset)
+	if err != nil {
+		return models.Asset{}, fmt.Errorf("failed to convert DB model to API model: %w", err)
+	}
+
+	return apiAsset, nil
+}
+
+func (t *AssetsTableHandler) CreateAsset(asset models.Asset) (models.Asset, error) {
+	// Check the user didn't provide an ID
+	if asset.Id != nil {
+		return models.Asset{}, &common.BadRequestError{
+			Reason: "can not specify id field when creating a new Asset",
+		}
+	}
+
+	// Check that assetInfo was provided by the user, it's a required field for an asset.
+	if asset.AssetInfo == nil {
+		return models.Asset{}, &common.BadRequestError{
+			Reason: "assetInfo is a required field",
+		}
+	}
+
+	// Generate a new UUID
+	asset.Id = utils.PointerTo(uuid.New().String())
+
+	// Initialise revision
+	asset.Revision = utils.PointerTo(1)
+
+	// TODO(sambetts) Lock the table here to prevent race conditions
+	// checking the uniqueness.
+	//
+	// We might also be able to do this without locking the table by doing
+	// a single query which includes the uniqueness check like:
+	//
+	// INSERT INTO scan_configs(data) SELECT * FROM (SELECT "<encoded json>") AS tmp WHERE NOT EXISTS (SELECT * FROM scan_configs WHERE JSON_EXTRACT(`Data`, '$.Name') = '<name from input>') LIMIT 1;
+	//
+	// This should return 0 affected fields if there is a conflicting
+	// record in the DB, and should be treated safely by the DB without
+	// locking the table.
```

Comment:
@sambetts-cisco just checking if this is going to impact anything

Reply:
This is an existing TODO, not something introduced by this PR (the file was just renamed, target changed to asset). We need to revisit some of these in the future.